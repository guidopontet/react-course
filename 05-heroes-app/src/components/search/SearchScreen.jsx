import { useLocation, useNavigate } from "react-router-dom";
import { useMemo } from "react";

import queryString from "query-string";

import useForm from "../../hooks/useForm";
import { getHeroeByName } from "../../services/heroes.service";
import { HeroCard } from "../hero/HeroCard";

export const SearchScreen = () => {
  const navigate = useNavigate();
  const location = useLocation();

  const { q: query = '' } = queryString.parse(location.search);

  const [{ searchText }, handleSearchTextChange] = useForm({ searchText: query});

  const handleSubmit = (e) => {
    e.preventDefault();

    navigate(`?q=${ searchText }`);
  }

  const heroesFiltered = useMemo(() => getHeroeByName(query), [ query ]);

  return (
    <>
      <h1>Búsquedas</h1>
      <hr />

      <div className="row">
        <div className="col-5">
          <h4>Buscar</h4>
          <hr />

          <form onSubmit={ handleSubmit }>
            <input type="text"
                   name="searchText"
                   placeholder="Buscar un héroe"
                   className="form-control"
                   autoComplete="off"
                   value={ searchText }
                   onChange={ handleSearchTextChange }/>

            <button type="submit"
                    className="btn btn-outline-primary mt-3 w-100">
              Buscar
            </button>
          </form>
        </div>

        <div className="col-7">
          <h4>Resultados</h4>
          <hr />

          {
            (query === '')
              ? <div className="alert alert-info">Buscar un héroe</div>
              : ( heroesFiltered.length === 0 )
                && <div className="alert alert-danger">No hay resultados: { query }</div>
          }

          {
            heroesFiltered.map((hero) => (
              <HeroCard key={ hero.id }
                        { ...hero }/>
            ))
          }
        </div>
      </div>
    </>
  )
}
