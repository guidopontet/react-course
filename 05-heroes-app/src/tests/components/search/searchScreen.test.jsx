import { mount } from "enzyme"
import { MemoryRouter } from "react-router-dom"

import { SearchScreen } from "../../../components/search/SearchScreen"

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}))

describe('SearchScreen', () => {
  test('should render with default values', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search']}>
        <SearchScreen />
      </MemoryRouter>
    )

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('input[name="searchText"]').prop('value')).toBe('');
  })

  test('should show batman with "batman" as seachText', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <SearchScreen />
      </MemoryRouter>
    )

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('input[name="searchText"]').prop('value')).toBe('batman');
  })

  test('should show error when no hero match', () => {
    const searchText = '1234';

    const wrapper = mount(
      <MemoryRouter initialEntries={[`/search?q=${searchText}`]}>
        <SearchScreen />
      </MemoryRouter>
    )

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.alert-danger').text()).toBe(`No hay resultados: ${searchText}`);
  })

  test('should call handleSubmit on search', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search']}>
        <SearchScreen />
      </MemoryRouter>
    )

    wrapper.find('input').simulate('change', { target: { name: 'searchText', value: 'batman' } });
    wrapper.find('form').simulate('submit', { preventDefault: () => {} });

    expect(mockNavigate).toHaveBeenCalledWith('?q=batman');
  })
})
