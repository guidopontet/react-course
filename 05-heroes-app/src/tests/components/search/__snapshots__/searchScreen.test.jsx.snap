// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchScreen should render with default values 1`] = `
Array [
  <h1>
    Búsquedas
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Buscar
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Buscar un héroe"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary mt-3 w-100"
          type="submit"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Buscar un héroe
      </div>
    </div>
  </div>,
]
`;

exports[`SearchScreen should show batman with "batman" as seachText 1`] = `
Array [
  <h1>
    Búsquedas
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Buscar
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Buscar un héroe"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-outline-primary mt-3 w-100"
          type="submit"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="col my-2 animate__animated animate__fadeIn"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-4"
            >
              <img
                alt="Batman"
                className="card-img"
                src="assets/img/dc-batman.jpg"
              />
            </div>
            <div
              className="col-md-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Batman
                </h5>
                <p
                  className="card-text"
                >
                  Bruce Wayne
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  >
                    Detective Comics #27
                  </small>
                </p>
                <a
                  href="/hero/dc-batman"
                  onClick={[Function]}
                >
                  <button
                    className="btn btn-primary"
                  >
                    Ver más
                  </button>
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`SearchScreen should show error when no hero match 1`] = `
Array [
  <h1>
    Búsquedas
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Buscar
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Buscar un héroe"
          type="text"
          value="1234"
        />
        <button
          className="btn btn-outline-primary mt-3 w-100"
          type="submit"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="alert alert-danger"
      >
        No hay resultados: 
        1234
      </div>
    </div>
  </div>,
]
`;
